packages/blocks/src/api/parser/apply-block-deprecated-versions.js

```diff
@@ -113,7 +113,11 @@ export function applyBlockDeprecatedVersions( block, rawBlock, blockType ) {
 
 		block = {
 			...block,
-			attributes: migratedAttributes,
+			attributes: getBlockAttributes(
```

Comment:
The problem with using `getBlockAttributes` here is that `migratedBlock.originalContent` is still the old block's content, so any blocks sourcing block attributes from the HTML will be incorrect.

Maybe we need a new function to skip those in this step.

Reply:
we're trying to source the new attributes here from the migrated block's new `save` output?

I'm missing something on what the problem is here, but it seems like we should be able to resolve it whatever it is

Reply:
this is all really confusing to me, and `applyBlockDeprecatedVersions` seems like it would benefit from drawing new terms and avoiding conflating names.

> migratedBlock.originalContent is still the old block's content, so any blocks sourcing block attributes from the HTML will be incorrect.

there could be a few ways this goes: the block's HTML doesn't contain the thing that the new attribute definitions are looking for, or it does, or it doesn't have the structure those attribute definitions are looking for.

 - for `source[src]` with `<img src="…">` we hopefully is `undefined` and an invalid parse.
 - for `img[src]` with `<img>` we get `''` which is a valid parse but wrong because it provides an empty string instead of an `undefined` value.
 - for `p:content` with `<p>This used to match everything.</p>` we get `'This used to match everything.'` which is a valid parse.

after creating `migratedBlock` we should be sourcing the _new_ attributes from the _old_ HTML, leaving us with missing or empty parameters that should be supplied by defaults if they aren't there. I'm reminded of the list block which at one point wiped out an entire list of save because the deprecation process returned an empty string instead of saying the block was invalid. I think in those cases changing `<ul>` to `<ult>` or some other typo was enough to trigger this flow.

we could also be hitting this backwards: HTML that was generated in a newer version of the block than we currently support. again, we could fail to parse structure and result in an empty value that wipes out content on save instead of preserving it.

maybe the most important piece is knowing if the original content was newer or older than the block in the editor, which actually is impossible due to the lack of versioning.

I _think_ that the idea in this function is to re-attempt to parse the original content using the newer attribute definitions, and I _think_ this PR is saying, "if after that re-parse we are missing attributes, supply the defaults." if that's the case it seems like we should call `addDefaultAttributes()` inside the deprecation loop and not after it, but I'm not sure that's still what we want.

 - `migratedBlock` sources attributes from new definitions
 - it's sent to `validateBlock`, which re-serializes the block with the newly-sourced attributes, then saves it, then compares if it matches the `originalContent`. it's it's valid it implies that the right data was there; probably the attribute definitions had a minor update.
 - if it's not validated it goes to the next deprecation in the list. this implies that loading and saving led to a different outcome; something is probably missing.
 - if it _is_ valid we look for a `migrate` function to transform it further. at this point we can read and write the block and get the same thing we started with.

the `validateBlock` function calls `save` with missing attributes, so it seems like we need to supply defaults to `migratedBlock` but also refrain from persisting them until deprecation/migration is over, since the next deprecation in the loop could supply a default that's missing from the current one.

```js
let migratedBlock = {
	...block,
	attributes: getBlockAttributes(
		deprecatedBlockType,
		block.originalContent,
		parsedAttributes
	),
};

const withDefaultAttributes = withDefaultsForMissingAttributes( deprecatedBlockType, migratedBlock );

let [ isValid ] = validateBlock( withDefaultAttributes, deprecatedBlockType );

if ( ! isValid ) {

}

migratedBlock = withDefaultAttributes;
```